### ISSUE TITLE
404 exception in docker and proxy server

### ISSUE THREAD

**Issue by mmikirtumov** [PiranhaCMS/piranha.core#1093]
Hi,

I am trying to run piranha using docker and in front of it nginx proxy server.

Docker image is mcr.microsoft.com/dotnet/core/aspnet:3.1.2

When I try to open piranha CMS it always sends 404 response. I tried to get some logs and here is the picture after any request

```info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.0 GET http://mydomain:443/cms/manager  
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.1 initialized 'Db' using provider 'Npgsql.EntityFrameworkCore.PostgreSQL' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (6ms) [Parameters=[@__siteId_0='?' (DbType = Guid), @__slug_1='?'], CommandType='Text', CommandTimeout='30']
      SELECT t."Id", t."ContentType", t."Created", t."IsHidden", t."LastModified", t."MetaDescription", t."MetaKeywords", t."NavigationTitle", t."OriginalPageId", t."PageTypeId", t."ParentId", t."Published", t."RedirectType", t."RedirectUrl", t."Route", t."SiteId", t."Slug", t."SortOrder", t."Title", t0."Id", t0."BlockId", t0."PageId", t0."ParentId", t0."SortOrder", t0."Id0", t0."CLRType", t0."Created", t0."IsReusable", t0."LastModified", t0."ParentId0", t0."Title", t0."Id1", t0."BlockId0", t0."CLRType0", t0."FieldId", t0."SortOrder0", t0."Value", p3."Id", p3."CLRType", p3."FieldId", p3."PageId", p3."RegionId", p3."SortOrder", p3."Value"
      FROM (
          SELECT p."Id", p."ContentType", p."Created", p."IsHidden", p."LastModified", p."MetaDescription", p."MetaKeywords", p."NavigationTitle", p."OriginalPageId", p."PageTypeId", p."ParentId", p."Published", p."RedirectType", p."RedirectUrl", p."Route", p."SiteId", p."Slug", p."SortOrder", p."Title"
          FROM "Piranha_Pages" AS p
          WHERE (p."SiteId" = @__siteId_0) AND (p."Slug" = @__slug_1)
          LIMIT 1
      ) AS t
      LEFT JOIN (
          SELECT p0."Id", p0."BlockId", p0."PageId", p0."ParentId", p0."SortOrder", p1."Id" AS "Id0", p1."CLRType", p1."Created", p1."IsReusable", p1."LastModified", p1."ParentId" AS "ParentId0", p1."Title", p2."Id" AS "Id1", p2."BlockId" AS "BlockId0", p2."CLRType" AS "CLRType0", p2."FieldId", p2."SortOrder" AS "SortOrder0", p2."Value"
          FROM "Piranha_PageBlocks" AS p0
          INNER JOIN "Piranha_Blocks" AS p1 ON p0."BlockId" = p1."Id"
          LEFT JOIN "Piranha_BlockFields" AS p2 ON p1."Id" = p2."BlockId"
      ) AS t0 ON t."Id" = t0."PageId"
      LEFT JOIN "Piranha_PageFields" AS p3 ON t."Id" = p3."PageId"
      ORDER BY t."Id", t0."Id", t0."Id0", t0."Id1", p3."Id"
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (2ms) [Parameters=[@__siteId_0='?' (DbType = Guid), @__slug_1='?'], CommandType='Text', CommandTimeout='30']
      SELECT t."Id", t."ContentType", t."Created", t."IsHidden", t."LastModified", t."MetaDescription", t."MetaKeywords", t."NavigationTitle", t."OriginalPageId", t."PageTypeId", t."ParentId", t."Published", t."RedirectType", t."RedirectUrl", t."Route", t."SiteId", t."Slug", t."SortOrder", t."Title", t0."Id", t0."BlockId", t0."PageId", t0."ParentId", t0."SortOrder", t0."Id0", t0."CLRType", t0."Created", t0."IsReusable", t0."LastModified", t0."ParentId0", t0."Title", t0."Id1", t0."BlockId0", t0."CLRType0", t0."FieldId", t0."SortOrder0", t0."Value", p3."Id", p3."CLRType", p3."FieldId", p3."PageId", p3."RegionId", p3."SortOrder", p3."Value"
      FROM (
          SELECT p."Id", p."ContentType", p."Created", p."IsHidden", p."LastModified", p."MetaDescription", p."MetaKeywords", p."NavigationTitle", p."OriginalPageId", p."PageTypeId", p."ParentId", p."Published", p."RedirectType", p."RedirectUrl", p."Route", p."SiteId", p."Slug", p."SortOrder", p."Title"
          FROM "Piranha_Pages" AS p
          WHERE (p."SiteId" = @__siteId_0) AND (p."Slug" = @__slug_1)
          LIMIT 1
      ) AS t
      LEFT JOIN (
          SELECT p0."Id", p0."BlockId", p0."PageId", p0."ParentId", p0."SortOrder", p1."Id" AS "Id0", p1."CLRType", p1."Created", p1."IsReusable", p1."LastModified", p1."ParentId" AS "ParentId0", p1."Title", p2."Id" AS "Id1", p2."BlockId" AS "BlockId0", p2."CLRType" AS "CLRType0", p2."FieldId", p2."SortOrder" AS "SortOrder0", p2."Value"
          FROM "Piranha_PageBlocks" AS p0
          INNER JOIN "Piranha_Blocks" AS p1 ON p0."BlockId" = p1."Id"
          LEFT JOIN "Piranha_BlockFields" AS p2 ON p1."Id" = p2."BlockId"
      ) AS t0 ON t."Id" = t0."PageId"
      LEFT JOIN "Piranha_PageFields" AS p3 ON t."Id" = p3."PageId"
      ORDER BY t."Id", t0."Id", t0."Id0", t0."Id1", p3."Id"
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 29.4576ms 404 ```

I think something related to database. I tried to find the reason and put more logs, but I couldn't find the place. 

Locally everything works fine.

**Can you please give some advice how can we handle this kind of problems.**
There is no information what is missing and what is wrong.

Thanks,
Mikayel

**Comment by tidyui** [comment]
It looks like you’re requesting the manager which should be visible even without data in the database. Is the project working locally before being packaged with docker?

**Comment by mmikirtumov** [comment]
Hi @tidyui ,

Thanks for reply.
Locally when we run it using Visual Studio or Rider it works and also logs I see the same request to database but after instead of 404 I see the following logs

```
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint '/Login'
info: Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker[3]
      Route matched with {page = "/Login", area = "Manager", action = "", controller = ""}. Executing page /Login
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.1 initialized 'IdentityPostgreSQLDb' using provider 'Npgsql.EntityFrameworkCore.PostgreSQL' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT EXISTS (SELECT 1 FROM pg_catalog.pg_class c JOIN pg_catalog.pg_namespace n ON n.oid=c.relnamespace WHERE c.relname='__EFMigrationsHistory');

```

**Comment by eric-wilson** [comment]
@mmikirtumov - I ran into some odd behaviors when running in docker w/ nginx too.  

The first issue I has was Nginx restarting constantly.  

While it's running with a docker-compose up (you may see the quick an occasional message of ngnix restarting) if you do a docker ps you may see this in the status of Nginx "Restarting (1) 2 seconds ago  "

It had to do with the names of the services not getting recognized w/in Nginx.  After I did a docker-compose build it restructured the docker cache and all was well.

The next issue I had was with some redirects going back to port 80.

I'm running it in docker with port mappings of Nginx on 8181:80, and the RazorWeb on 5000 (all in the same docker network) so to access it locally I go to http://localhost:8181 but there are times on redirects that it goes back to 80 e.g. http://localhost, which of course gives me a 404.  Perhaps you are experiencing something similar. So far this has only happened on manager redirects (I think).

If this doesn't help I can post my dockerfile(s) and compose files if want to see them.

**Comment by tidyui** [comment]
Hi @mmikirtumov! Did the comment from @eric-wilson solve your issues?

**Comment by eric-wilson** [comment]
@mmikirtumov , one other thing I noticed.  It looks like you are using the force SSL and the manager is under a different directory of `cms`.  depending on how you have it configured, those two items may cause an issue.

```
Request starting HTTP/1.0 GET http://mydomain:443/cms/manager 
```

Here's a working example of Nginx w/ Piranha:

https://github.com/geekcafe/piranha-docker

**Comment by mmikirtumov** [comment]
Hi @eric-wilson 

Thank you for your suggestions and help.

Actually the problem solved when I changed the configuration of nginx. Proxy server configured in this way. All requests which is coming with cms should proxy to PiranhaCMS docker. 

``
location /cms {
			proxy_pass http://localhost:5002/;
		}
``
it always throw the exception as described in this issue.

I just changed it to 
``
location /{
			proxy_pass http://localhost:5002/;
		}
``
and now everything works fine.

But still have a problem with /cms.
